src/primitives/sinsemilla.rs

```diff
@@ -0,0 +1,97 @@
+//! The Sinsemilla hash function and commitment scheme.
+
+use std::iter;
+
+use group::{Curve, Group};
+use halo2::{
+    arithmetic::{CurveAffine, CurveExt},
+    pasta::pallas,
+};
+
+const GROUP_HASH_Q: &str = "z.cash:SinsemillaQ";
+const GROUP_HASH_S: &str = "z.cash:SinsemillaS";
+
+const K: usize = 10;
+const C: usize = 253;
+
+fn lebs2ip_32(bits: &[bool]) -> u32 {
+    bits.iter()
+        .enumerate()
+        .fold(0u32, |acc, (i, b)| acc + if *b { 1 << i } else { 0 })
+}
+
+/// Hash extractor for Pallas, from [§ 5.4.8.7].
+///
+/// [§ 5.4.8.7]: https://zips.z.cash/protocol/orchard.pdf#concreteextractorpallas
+fn extract(point: &pallas::Point) -> pallas::Base {
```

Comment:
Same here, should probably be `pallas::Affine`; `Extract_P` is defined as having a `P` element input which is an affine point:
![image](https://user-images.githubusercontent.com/552961/109359536-a3c2bf80-7853-11eb-926c-4dc6fc3a2172.png)
![image](https://user-images.githubusercontent.com/552961/109359989-6ad71a80-7854-11eb-8c75-0387435c33c7.png)

Style opinion, as this is an open PR vs needing to open a new thing in `halo2`: naming all point types as explicitly *Point, such as `AffinePoint`, `ProjectivePoint`, etc, which can improve readability and human distinguishability of types. The [`jubjub` crate](https://docs.rs/jubjub/0.6.0/jubjub/) basically does this. 

I know in `halo2` a lot of these are generated by macros but the [names are chosen on macro invocation](https://github.com/zcash/halo2/blob/63083e418cdd836e0056dd55f6498ccf185af83f/src/pasta/curves.rs#L908) and renamed and exported in the [`pallas` module](https://github.com/zcash/halo2/blob/main/src/pasta/pallas.rs#L15).  This is of course not limited to or blocking on this PR but is a kind of naming / typing style thing we zcash engineers can discuss, if desired.

Reply:
Having this take `pallas::Affine` pushes a `point.to_affine()` call onto every caller, instead of locating them all inside this function. There are only a couple of callers though, so I don't overly mind.

Re: naming of `pallas::Point`, I opted in the `halo2` crate to leave the point representation out of the re-exported name for overall readability (and also because as above, it's irrelevant to the consumer). What mattered more to me was that I could write `pallas::Point` _everywhere_, so that it was clear that this is a Pallas point (rather than Vesta or some other curve); `pallas::ProjectivePoint` is almost twice as long.

`jubjub` is a slightly different case, because we need to represent both the curve type _and_ whether the point is constrained to the prime-order subgroup, so we already need to include more information in the name (in hindsight `jubjub::FullPoint` would have been a better name for this purpose than `jubjub::ExtendedPoint`). Pallas and Vesta are prime-order curves and don't have this problem.

Reply:
`pallas::ProjectivePoint` may be longer but it's not super long, and since there are multiple point types, vs a 'Point' and an 'Affine', it would be nice in terms of having the code document itself. Luckily Rust compiles away token names so we need not be precious on name lengths, as long we we don't turn into Java (😅 )

Reply:
I'm only weakly against a rename (I like Dodger's "weakly" vs "strongly" for indicating preference), so I'll defer to others on this.

Reply:
I like the current short names.